prusti-viper/src/encoder/mir_encoder/mod.rs

```diff
@@ -20,13 +20,93 @@ use rustc_middle::{mir, ty};
-use std::collections::HashMap;
+use std::{
+    collections::HashMap,
+    fmt::Display,
+};
 use prusti_interface::environment::mir_utils::MirPlace;
 use downcast_detector::detect_downcasts;
 
 pub static PRECONDITION_LABEL: &'static str = "pre";
 pub static WAND_LHS_LABEL: &'static str = "lhs";
 
+/// Result of encoding a place
+#[derive(Debug, PartialEq, Eq, Hash, Clone)]
+pub enum PlaceEncoding {
+    /// Just an expression, the most common case.
+    Expr(vir::Expr),
+    /// Field access expression
+    FieldAccess {
+        base: Box<PlaceEncoding>,
+        field: vir::Field,
+    },
+    /// Array access expression
+    ArrayAccess {
+        base: Box<PlaceEncoding>,
+        index: vir::Expr,
+        // TODO: prusti-common::src::vir::optimizations::purification has fn translate_type
+        array_elem_ty: vir::Type,
+        array_len: usize,
+    },
+    Variant {
+        base: Box<PlaceEncoding>,
+        field: vir::Field,
+    }
+}
+
+impl PlaceEncoding {
+    pub fn try_into_expr(self) -> EncodingResult<vir::Expr> {
+        match self {
+            PlaceEncoding::Expr(e) => Ok(e),
+            PlaceEncoding::FieldAccess { base, field } => {
+                Ok(base.try_into_expr()?.field(field))
+            },
+            PlaceEncoding::Variant { base, field } => {
+                Ok(vir::Expr::Variant(box base.try_into_expr()?, field, vir::Position::default()))
+            },
+            other => Err(EncodingError::internal(format!("PlaceEncoding::try_into_expr called on non-expr '{:?}'", other))),
+        }
+    }
+
+    pub fn field(self, field: vir::Field) -> Self {
+        PlaceEncoding::FieldAccess { base: box self, field }
+    }
+
+    pub fn get_type(&self) -> &vir::Type {
+        match self {
+            PlaceEncoding::Expr(ref e) => e.get_type(),
+            PlaceEncoding::FieldAccess { ref field, .. } => &field.typ,
+            PlaceEncoding::ArrayAccess { ref array_elem_ty, .. } => array_elem_ty,
+            PlaceEncoding::Variant { ref base, ref field } => &field.typ,
+        }
+    }
+
+    pub fn variant(self, index: &str) -> Self {
+        // TODO: somewhat duplicate from vir::Expr::variant()
+        let field_name = format!("enum_{}", index);
+        let field = vir::Field::new(field_name, self.get_type().clone().variant(index));
+
+        PlaceEncoding::Variant { base: box self, field }
+    }
+}
+
+impl From<vir::Expr> for PlaceEncoding {
+    fn from(e: vir::Expr) -> Self {
+        PlaceEncoding::Expr(e)
+    }
+}
+
+impl Display for PlaceEncoding {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PlaceEncoding::Expr(e) => write!(f, "{}", e),
+            PlaceEncoding::FieldAccess { base, field } => write!(f, "{}.{}", base, field),
+            PlaceEncoding::ArrayAccess { base, index, .. } => write!(f, "{}[{}]", base, index),
+            PlaceEncoding::Variant { base, field } => write!(f, "{}[{}]", base, field),
```

Comment:
What about `"({} as {})"` to distinguish `Variant` from `ArrayAccess`?

Reply:
so somewhat like the mir printing does? this is from the encoding of `prusti-tests/tests/verify/pass/enums/basic.rs` (file `log/vir_program_before_foldunfold/basic.rs.vir`, line 386 here):
```
    // [mir] _6 = &'_ mut ((_2 as V2).0: u32)
    _6 := builtin$havoc_ref()
    inhale acc(_6.val_ref, write)
    _6.val_ref := mut borrow _2[enum_V2].f$0 // L0
```

so i'd probably make a change to make the last line `_6.val_ref := mut borrow (_2 as enum_V2).f$0` instead? that log/ stuff is never read back anywhere, right? so i'd only have to change what is printed, no parser or anything else?

i'd do that in a separate PR probably, this one feels heavy already, and i quite like the "small PRs, easier/quicker reviews, faster small-step progress" approach.

Reply:
I didn't notice that `field` is a string and not a number, so it's impossible to confuse the two variants. It's fine as you did.